Add API test verifying that session state restore works with site isolation

beidson opened the pull request:
#### 346b1b2b7b56e3871b456e9093016265dfd90ff4
<pre>
Add API test verifying that session state restore works with site isolation
<a href="https://rdar.apple.com/162727442">rdar://162727442</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300843">https://bugs.webkit.org/show_bug.cgi?id=300843</a>

Reviewed by Sihui Liu.

We can&apos;t verify full fidelity *session save* with site isolation on due to continued work
needed on the UI process back/forward list.

But we can verify session restore works when the restored blob comes from a non-isolated view,
and verify that the resulting session save matches what was restored from the non-isolated view.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/Shared/SessionState.cpp:
(WebKit::SessionState::isEqualForTesting const):
(WebKit::BackForwardListState::isEqualForTesting const):
(WebKit::FrameState::isEqualForTesting const):
* Source/WebKit/Shared/SessionState.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKSessionState.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKSessionState.mm:
(-[_WKSessionState isEqualForTesting:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::siteIsolatedViewAndDelegate):
(TestWebKitAPI::viewAndDelegate):
(TestWebKitAPI::TEST(SiteIsolation, ValidateSessionRestoreWithoutNavigating)):

Canonical link: <a href="https://commits.webkit.org/301720@main">https://commits.webkit.org/301720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5d3040d366597babf7a5c6a464b4cf9894ed7ac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126742 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46384 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37351 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133699 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78376 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/17aaeb1d-4dd4-414e-ae5e-ae1965c72701) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54921 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96443 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129690 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37613 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113358 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76967 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36496 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31542 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77092 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107431 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31841 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136276 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41104 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104955 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53914 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109718 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104658 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26709 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50154 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28493 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50871 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53348 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59146 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52607 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55945 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54356 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKSessionState.h:
```diff
@@ -31,6 +31,7 @@ WK_CLASS_AVAILABLE(macos(10.10), ios(8.0))
 @interface _WKSessionState : NSObject
 
 - (instancetype)initWithData:(NSData *)data;
+- (BOOL)isEqualForTesting:(_WKSessionState *)other;
```
This should probably have availability macros.

beidson committed:
Add API test verifying that session state restore works with site isolation
rdar://162727442
https://bugs.webkit.org/show_bug.cgi?id=300843

Reviewed by Sihui Liu.

We can't verify full fidelity *session save* with site isolation on due to continued work
needed on the UI process back/forward list.

But we can verify session restore works when the restored blob comes from a non-isolated view,
and verify that the resulting session save matches what was restored from the non-isolated view.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/Shared/SessionState.cpp:
(WebKit::SessionState::isEqualForTesting const):
(WebKit::BackForwardListState::isEqualForTesting const):
(WebKit::FrameState::isEqualForTesting const):
* Source/WebKit/Shared/SessionState.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKSessionState.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKSessionState.mm:
(-[_WKSessionState isEqualForTesting:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::siteIsolatedViewAndDelegate):
(TestWebKitAPI::viewAndDelegate):
(TestWebKitAPI::TEST(SiteIsolation, ValidateSessionRestoreWithoutNavigating)):

Canonical link: https://commits.webkit.org/301720@main
